design-proposals/klip-5-internal-schemas-for-KSQL-engine.md

```diff
@@ -0,0 +1,191 @@
+KSQL engine would build the same internal schema for a given external schema everytime. As an example,
+ let’s consider we use the index of the field in the schema as a prefix to a constant string value to
+ build the internal schema field names. Consider the previous example where our external schema has
+ three fields as the following:
+
+```sql
+(“@ID” BIGINT, “@NAME” STRING, “MESSAGE.AMOUNT” DOUBLE)
+```
+
+Assuming the constant string prefix to build the internal schema fields is “COL_”, the internal
+schema for the above schema will be as the following:
+
+```sql
+(COL_0 BIGINT, COL_1 STRING, COL_2 DOUBLE)
+```
+
+Note that the following one to one mapping will exist between the fields from the external schema and the fields from the internal schema:
+
+* “@ID”                            <==>   COL_0
+* “@NAME”                     <==>   COL_1
+* “MESSAGE.AMOUNT” <==>   COL_2
+
+In the case of nested fields, we can apply the same approach to the nested fields inside ARRAY, MAP
+and STRUCT types.
+
+Now that we have the internal schema built, the engine will rewrite the queries by replacing the
+identifiers that refer to the fields in the external schema with their corresponding field name in
+the internal schema.
+
+As an example, consider the following query again:
+
+```sql
+CREATE STREAM bar AS
+SELECT “@ID”, lowercase(“@NAME”) AS “@NAME”, “MESSAGE.AMOUNT” * 100 AS “AMOUNT.BY.100”
+FROM foo
+WHERE “MESSAGE.AMOUNT” < 1;
+```
+
+With the above mapping the engine will write this query as the following:
+
+```sql
+
+CREATE STREAM bar AS
+SELECT FOO_COL_1 AS “@ID”, lowercase(FOO_COL_1) AS “@NAME”, FOO_COL_2 * 100 AS “AMOUNT.BY.100”
+FROM foo
+WHERE FOO_COL_2 < 1;
+```
+
+KSQL engine will be able to run the above query since all identifiers conform to the identifier
+naming requirements in KSQL. Note that KSQL already appends the source name to the field name
+automatically which will eliminate the ambiguity in case of having multiple sources (such as Join queries)
+ with the same field names.
+
+The engine will use the generated internal schema to build the streams application. The deserializers
+ from the source stream or table will use the external schema to fetch the fields and build the
+ GenericRow object, but beyond this point the internal schema will be used to process the records
+ for the queries.
+
+Also note that the result schema will be generated from the external schemas which means if we select
+any field from the external schema in the SELECT expressions and user does not specify an alias for
+the result column, the engine will use the same name from the external schema for the corresponding
+field in the result schema.
+
+Note that all the KSQL error or log messages should still use the external schema names since user
+does not know about the internal schemas.
+
+## Test Plan
+
+Tests for supporting quoted identifiers will be added.
```

Comment:
With the ability to quote identifiers, it would now be possible for the user to create field names that are not valid for the serialization format being used.  We should ensure this results in an error when the statement is submitted, rather than just result in runtime serde errors.

Reply:
Exactly, the engine should enforce the requirements of the sink format. For instance, if the sink is AVRO and there is a fields name that is not valid for avro the engine should enforce this in compile time, not runtime. I will clarify this in the KLIP.